### DISCUSSION TITLE
[CI Pipeline] Error "yaml: invalid map key" during mlp kustomize

### DISCUSSION THREAD

**Discussion by seraph-eeno** [mia-platform/community#143] [Q&A]
Hi everyone,

we're building up a new Console project dedicated only to Fast Data management.

We've settled all the settings and microservices and we wanted to deploy in order to test the configuration. Unluckily, the CI pipeline on GitLab fails due to the following error during the **mlp kustomize** phase:

**`map[string]interface {}(nil): yaml: invalid map key: map[string]interface {}{"SUSPEND_CSV_IMPORT":""}`**.

Here's all the listing (I've used asterisks to cover some info):
```
Running with gitlab-runner 15.4.0 (43b2dc3d)
  on deploy-europe-west1-c *****
  feature flags: FF_USE_FASTZIP:true
Preparing the "docker+machine" executor
00:03
Using Docker executor with image miaplatform/mlp:1.1.0 ...
Pulling docker image miaplatform/mlp:1.1.0 ...
Using docker image sha256:***** for miaplatform/mlp:1.1.0 with digest miaplatform/mlp@sha256:***** ...
Preparing environment
00:00
Running on runner-vzisqcy2-project-51950-concurrent-0 via runner-vzisqcy2-autoscale-1671468920-3ff8d605...
Getting source from Git repository
00:01
Fetching changes with git depth set to 1...
Reinitialized existing Git repository in /builds/*****/fast-data/Configurations/.git/
Checking out 9af5ad3f as master...
Removing interpolated-files/
Removing overlays/DEV/nexus-gcloud.secret.yaml
Updating/initializing submodules with git depth set to 1...
Executing "step_script" stage of the job script
00:02
Using docker image sha256:***** for miaplatform/mlp:1.1.0 with digest miaplatform/mlp@sha256:***** ...
$ export RELEASE_DATE="$(date -I'seconds' -u)"
$ mkdir "${DESTINATION_PATH}"
$ echo "${KUBE_CA_PEM}" > /tmp/kube-ca.pem
$ test -f "${VARIABLES_FILE}" && set -a && source "${VARIABLES_FILE}"
$ test -f "${GENERATE_FILE}" && mlp generate -c "${GENERATE_FILE}" -e "${ENVIRONMENT_PREFIX}" -e "${ENVIRONMENT_VARIABLES_PREFIX}" -o "${OVERLAY_PATH}"
$ mlp hydrate "${BASE_PATH}" "${OVERLAY_PATH}"
$ mlp kustomize "${OVERLAY_PATH}" -o "${DESTINATION_PATH}/kustomize-output.yaml"
map[string]interface {}(nil): yaml: invalid map key: map[string]interface {}{"SUSPEND_CSV_IMPORT":""}
Uploading artifacts for failed job
00:01
Uploading artifacts...
./interpolated-files: found 1 matching files and directories 
Uploading artifacts as "archive" to coordinator... 201 Created  id=1232352 responseStatus=201 Created token=xY6cPxNs
Cleaning up project directory and file based variables
00:01
ERROR: Job failed: exit code 1
```
`SUSPEND_CSV_IMPORT` is set in the Public Variables section and is used in a CronJob configuration for the `suspend` property of its yml file.
Things get bizarre now: I tried not to use that variable anymore and delete it from the configurations in Console. Checking on the git changes from the commit, deletions seemed effective to me. Deploying once again, the CI pipeline showed the same error on the same variable.

Yesterday, the same error was launched for another variable used to configure the number of replicas for a service, but deleting it moved the error to the variable `SUSPEND_CSV_IMPORT`.

Some little details on our Configurations project below.
Our repo follows this structure:
```
.
├── configuration
│   ├── kustomization.yaml
│   └── %resourceName%.yaml
└── overlays
    └── %envId%
        ├── kustomization.yaml
        ├── variables.env
```

Referring to the **DEV** stage pipeline, `%envId%` is being set to `DEV`.

If other details are requested, I will provide them.

**Comment by davidebianchi** [comment]
Inside the variables.env file, there is the env var? Is it suspend of the CronJobs? suspend is a boolean value inside CronJobs, could be this. Can you try to remove the interpolation from the CronJob?

I see that you use kustomize as project structure, you can use some kustomize patch to address this specific behavior.

**Reply by seraph-eeno** [reply]
Hi Davide, thanks for answering. Yes, the var is present in variables.env.
Before opening this discussion, I had already tried using the suspend without interpolation, therefore I used the plain boolean value in the CronJobs yml file; the error persisted.
Then I refined the test deleting the public variable entirely from the corresponding section in the Console and still nothing changed.

I will try adding a patch adn check again, though I've nevere used patch for kubernetes.

**Reply by fabio-percivaldi** [reply]
Hi Davide, we had the same exact problem with also the variable used for the number of replicas inside a microservice, i don't think it is closely related to the CronJobs, i didn't understand why public variables are not working with kustomize. 

By the way we inherited the kustomize structure from the common project template, it wasn't exactly a conscious choice.

**Reply by davidebianchi** [reply]
The support for the public variables is the same for both the project structure, so I don't think the two things are related. 
I have not understand what are you tried, have you removed the variables both from the usage and from the variables definition?
